hphp/util/asm-x64-intelxed.h

```diff
@@ -0,0 +1,1066 @@
+  }
+
+  ALWAYS_INLINE
+  uint32_t xedEmit(xed_iclass_enum_t instr,
+                   const xed_encoder_operand_t& op_1,
+                   const xed_encoder_operand_t& op_2,
+                   xed_uint_t effOperandSizeBits,
+                   CodeAddress destination) {
+    return xedEmitImpl(instr, destination,
+                        [&](xed_encoder_instruction_t* i) {
+                          xed_inst2(i, s_xedState, instr,
+                                    effOperandSizeBits, op_1, op_2);
+                        });
+  }
+
+  ALWAYS_INLINE
+  void xedEmit(xed_iclass_enum_t instr,
+               const xed_encoder_operand_t& op_1,
+               const xed_encoder_operand_t& op_2,
+               xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit(instr, op_1, op_2, effOperandSizeBits,
+                                   dest()));
+  }
+
+  ALWAYS_INLINE
+  uint32_t xedEmit(xed_iclass_enum_t instr,
+                   const xed_encoder_operand_t& op_1,
+                   const xed_encoder_operand_t& op_2,
+                   const xed_encoder_operand_t& op_3,
+                   xed_uint_t effOperandSizeBits,
+                   CodeAddress destination) {
+    return xedEmitImpl(instr, destination,
+                        [&](xed_encoder_instruction_t* i) {
+                          xed_inst3(i, s_xedState, instr,
+                                    effOperandSizeBits, op_1, op_2, op_3);
+                        });
+  }
+
+  ALWAYS_INLINE
+  void xedEmit(xed_iclass_enum_t instr,
+               const xed_encoder_operand_t& op_1,
+               const xed_encoder_operand_t& op_2,
+               const xed_encoder_operand_t& op_3,
+               xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit(instr, op_1, op_2, op_3, effOperandSizeBits,
+                                   dest()));
+  }
+
+public:
+#define INT3_SIZE   (sz::byte)
+#define UD2_SIZE    (sz::word)
+
+  void emitInt3s(int n) {
+    static uint8_t instr = 0;
+    if (n == 0) return;
+    if (instr == 0) {
+      xedEmit(XED_ICLASS_INT3, sz::byte, &instr);
+    }
+    for (auto i = 0; i < n; ++i) {
+      byte(instr);
+    }
+  }
+
+  void emitNop(int n) {
+    if (n == 0) return;
+    static const xed_iclass_enum_t nops[] = {
+      XED_ICLASS_INVALID,
+      XED_ICLASS_NOP,
+      XED_ICLASS_NOP2,
+      XED_ICLASS_NOP3,
+      XED_ICLASS_NOP4,
+      XED_ICLASS_NOP5,
+      XED_ICLASS_NOP6,
+      XED_ICLASS_NOP7,
+      XED_ICLASS_NOP8,
+      XED_ICLASS_NOP9,
+    };
+    // While n >= 9, emit 9 byte NOPs
+    while (n >= 9) {
+      xedInstr(XED_ICLASS_NOP9, 0);
+      n -= 9;
+    }
+    // Emit remaining NOPs (if any)
+    if (n) {
+      xedInstr(nops[n], 0);
+    }
+  }
+
+  void pad() {
+    auto remaining = available();
+    static uint8_t padInstructs[UD2_SIZE + INT3_SIZE] = {0};
+    if (padInstructs[0] == 0) {
+      xedEmit(XED_ICLASS_UD2, 0, padInstructs);
```

Comment:
Same comment about thread safety as line 557.

Reply:
Addressed in https://github.com/facebook/hhvm/pull/8164/commits/ec24a9fdfb463fda92b86f6ac4e1061488fa4d06